unstructured/ingest/connector/pinecone.py

```diff
@@ -0,0 +1,143 @@
+
+
+@dataclass
+class PineconeWriteConfig(IngestDocSessionHandleMixin, WriteConfig):
+    connector_config: SimplePineconeConfig
+    batch_size: int = 50
+    num_processes: int = 1
+
+
+@dataclass
+class PineconeDestinationConnector(BaseDestinationConnector):
+    write_config: PineconeWriteConfig
+    connector_config: SimplePineconeConfig
+
+    def initialize(self):
+        pass
+
+    def check_connection(self):
+        create_pinecone_object(
+            self.connector_config.api_key,
+            self.connector_config.index_name,
+            self.connector_config.environment,
+        )
+
+    @DestinationConnectionError.wrap
+    @requires_dependencies(["pinecone"], extras="pinecone")
+    def upsert_batch(self, batch):
+        import pinecone.core.client.exceptions
+
+        self.write_config.global_session()
+
+        index = self.write_config.session_handle.service
+        try:
+            response = index.upsert(batch)
+        except pinecone.core.client.exceptions.ApiException as api_error:
+            raise WriteError(f"http error: {api_error}") from api_error
+        logger.debug(f"results: {response}")
+
+    def write_dict(self, *args, dict_list: t.List[t.Dict[str, t.Any]], **kwargs) -> None:
+        logger.info(
+            f"Upserting {len(dict_list)} elements to destination "
+            f"index at {self.connector_config.index_name}",
+        )
+
+        pinecone_batch_size = self.write_config.batch_size
+
+        logger.info(f"using {self.write_config.num_processes} processes to upload")
+        if self.write_config.num_processes == 1:
+            for i in range(0, len(dict_list), pinecone_batch_size):
+                self.upsert_batch(dict_list[i : i + pinecone_batch_size])  # noqa: E203
+
+        else:
+            with mp.Pool(
+                processes=self.write_config.num_processes,
+            ) as pool:
+                pool.map(
+                    self.upsert_batch,
+                    [
+                        dict_list[i : i + pinecone_batch_size]  # noqa: E203
+                        for i in range(0, len(dict_list), pinecone_batch_size)
+                    ],  # noqa: E203
+                )
+
+    def write(self, docs: t.List[BaseIngestDoc]) -> None:
+        dict_list: t.List[t.Dict[str, t.Any]] = []
+        for doc in docs:
+            local_path = doc._output_filename
+            with open(local_path) as json_file:
+                dict_content = json.load(json_file)
+
+                # assign element_id to "id", embeddings to "values", and other fields to "metadata"
+                dict_content = [
+                    # While flatten_dict enables indexing on various fields,
+                    # element_serialized enables easily reloading the element object to memory.
+                    # element_serialized is formed without text/embeddings to avoid data bloating.
+                    {
+                        "id": element.pop("element_id", None),
```

Comment:
Setting the id, which I assume needs to be a unique identifier, is a bit of an issue using `element_id`. `element_id` is derived from the text and nothing else, meaning one element with text `hello world` but different metadata (such as location on the page) will have the same `element_id` as another element with the same text but exists elsewhere on the page. I think that needs to be addressed in the core library, but for now it prevents us from using it as the unique identified in a database because it will cause those elements to overwrite themselves and only one will exist. For now I used a random UUID generated for the id in other connectors if one isn't automatically generated for you by the provider.

Reply:
Agreed; I now applied the same approach as in azure cognitive search, and addressed with: [unique ids for pinecone entries rather than using element ids](https://github.com/Unstructured-IO/unstructured/pull/1774/commits/e353c6b2495539883f5796949ca6c51de9db9109)